### ISSUE TITLE
[ECS/Fargate] [request]: Allow stopTimeout to be configured for ondemand tasks

### ISSUE THREAD

**Issue by bishwa-tidal** [aws/containers-roadmap#1020]
Based on this comment: https://github.com/spring-projects/spring-boot/issues/4657#issuecomment-161354811, if we were to implement graceful termination on the application side, it would really help if `stopTimeout` would allowed to be configurable - at least for on-demand fargate tasks.

Currently the max value is `120s` which is not sufficient for all usecases.
 

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Tell us about your request**
What do you want us to build?

**Which service(s) is this request for?**
This could be Fargate, ECS

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
What outcome are you trying to achieve, ultimately, and why is it hard/impossible to do right now? What is the impact of not having this problem solved? The more details you can provide, the better we'll be able to understand and solve the problem.

**Are you currently working around this issue?**
How are you currently solving this problem?

**Additional context**
Anything else we should know?

**Attachments**
If you think you might have additional information that you'd like to include via an attachment, please do - we'll take a look. (Remember to remove any personally-identifiable information.)

**Comment by matteomazza91** [comment]
Implementing graceful termination on the application side we are facing the same issue.  
We have an ECS service with an ALB, Fargate tasks and autoscaling that increase and decrease the desiredCount.

When an ECS task receive SIGTERM (docker stop), it should get a chance to complete the ongoing work before being forcibly killed.  
Unfortunately for our use case a maximum value of `120s` for `stopTimeout` is not enough.

**Comment by bryanculbertson** [comment]
I run long running tasks on ECS and would like to run them in fargate. I would like for them to have a chance to finish which would mean a multi-hour stop timeout.

**Comment by mysticmoron** [comment]
The 2 minute limit is too short for many use-cases.
With this limit in place, AWS fargate auto-scaling is rendered useless for any system that cares about graceful shutdown.

**Comment by PatrickRice-KSC** [comment]
Agreed with other posters. We have long running Fargate tasks that listen to a queue for picking up tasks, and tasks have a potentially unbounded processing time. Yes, they'll pick up the message again from the queue after timeout if the tasks dies (once visibility expires), but I'd prefer that the task be allowed to finish and gracefully shut down.

With a 2 minute limit, we've been forced to abandon AWS's default auto scaling, and create a lambda that checks for scaling every minute, and sends a http request to containers to shut down instead of using sigterm.

**Comment by marc-guenther** [comment]
Same problem here. This renders Fargate useless for our application. Seems we need to use EC2 instances, where this limit does not exist?

We are paying for running containers even whey they are shutting down, why cannot we set the stopTimeout to whatever value we like?

**Comment by GytisZ** [comment]
Encountering a similar issue. We have tasks that we'd like to drain / shutdown slower than in two minutes. Currently that means that we're migrating from ECS Services and Autoscaling and will need to manually manage the tasks both during deployments and autoscaling. Using the RunTask and some scripts to hold it all together.

Being able to extend the StopTimeout to 1-2 hours is the only reason why the current setup doesn't work for us.

**Comment by keirwhitlock** [comment]
Bumping this! - Such a needed feature.

**Comment by satya-500** [comment]
StopTimeout should be set by users as much they want.

**Comment by maddipati-srinivas** [comment]
StopTimeout - max can be only 2 mints for SIGTERM. For more details go through the this [link](https://aws.amazon.com/blogs/containers/graceful-shutdowns-with-ecs/). But this solution will not work for stateful operations as it's depends on our business logic.

May I know when we can expect full pledge solution from AWS?

**Comment by mdomsch-seczetta** [comment]
We moved our (long-running) batch processing application from ECS on Fargate to ECS on EC2 so that we could manage the termination behavior and extend it as long as necessary in order to properly let our batch jobs complete and drain safely without loss of work.  2 minutes is woefully insufficient. However, this has lead to significantly increased DataDog monitoring costs (from ~$1.40/task to $56/task), which cannot be borne in our budget.  We'd be happy to keep the tasks on Margate, if the StopTimeout could be extended as long as necessary.

**Comment by craigify** [comment]
Yes we have the exact same problem. I am not able to use ECS for one of our major applications because I need to allow a Fargate instance much more time than 2 minutes to shut down.

**Comment by ADrejta** [comment]
+1 on this. The 2 minute maximum timeout is quite low for graceful shutdowns of longer running jobs. We either have to implement a retry mechanism of jobs that might be lost or completely switch to running ECS in EC2 mode where that number can be set much higher.

**Comment by cj-arnas-navasinskas** [comment]
+1, ECS stopTimeout limitations are impacting graceful shutdowns.

**Comment by Abdelwahab313** [comment]
Hi @AbhishekNautiyal 
Is there any update on resolving this issue?

I'm currently designing an architecture for deploying multiple Delayed Job background workers and aiming to run multiple workers within a single container. However, I believe the Task Protection API doesn’t fit my use case for the following reasons:

Maintaining the state of parallel workers is complex—determining which workers should request protection and which should release it adds operational overhead.

My current approach relies on the fact that if a deployment blocks a claim response, then tasks are not accepted, which effectively pauses execution during scale-in events or new deployments. This could lead to extended downtime, as all tasks need to release protection before scaling operations can proceed.

Given these limitations with Task Protection on Fargate, I am considering switching to ECS running on EC2. While Fargate initially seemed promising, this limitation significantly impacts use cases involving autoscaling for multi-threaded workers. This restriction may be blocking a key segment of users who depend on such workloads.

Could you share whether this issue is on the roadmap or if it is considered a low-priority enhancement?

Additionally, I understand there may be a two-minute limit on Fargate Spot claim protection—could that be a contributing factor to this limitation?

Looking forward to your insights!

**Comment by krailler** [comment]
+1, we need more than 2 minutes

**Comment by wassim-alma** [comment]
+1, this is so painful, this is not allowing long running Celery tasks to finish gracefully, we can't do autoscaling and deployment could lead to lost tasks!

**Comment by TatevHunanyan-BM** [comment]
+1

**Comment by MustaphaAwwal** [comment]
+1

**Comment by yokawasa** [comment]
+1

**Comment by purplepenguin2** [comment]
@AbhishekNautiyal Can you provide any update here? The `120` second maximum for `stopTimeout` on ECS Fargate is insanely limiting and causes a lot of problems for certain use cases.

**Comment by susanwei126** [comment]
+++++++1111111

**Comment by raylillywhite** [comment]
+1

**Comment by holma91** [comment]
+1

**Comment by sjgutta** [comment]
+1

**Comment by JimNero009** [comment]
I am also in the unenviable position now where this 2 minutes might look to be insufficient. Please can you let this be increased to an arbitrary value, or at least within a more reasonable limit?

**Comment by ttahir86** [comment]
+1

**Comment by mattmilesi** [comment]
+1. To terminate tasks properly, 120 seconds is definitely too short for many use cases.
